src/Geometry/Primitives/AsphericSurface.jl

```diff
@@ -0,0 +1,341 @@
+# MIT license
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# See LICENSE in the project root for full license information.
+
+
+
+#########################################################################################################
+
+
+"""
+
+```julia
+AsphericSurface(semidiameter; radius, conic, aspherics=nothing, normradius = semidiameter)
+```
+
+Surface incorporating an aspheric polynomial - radius, conic and aspherics are defined relative to absolute semi-diameter,.
+`T` is the datatype, `N` is the dimensionality,  and `Q` is the number of aspheric terms. 
+
+The surface is centered at the origin and treated as being the cap of an infinite cylinder, thus creating a true half-space.
+Outside of `0 <= ρ <= 1` the height of the surface is not necessarily well defined, so NaN may be returned.
+
+`aspherics`  aspherics should be vectors containing tuples of the form (i, v) where i is the polynomial power of the aspheric term
+
+The sag is defined by the equation
+
+```math
+z(r,\\phi) = \\frac{cr^2}{1 + \\sqrt{1 - (1+k)c^2r^2}} + \\sum_{i}^{Q}\\alpha_ir^{i}
+```
+
+where ``\\rho = \\frac{r}{\\texttt{normradius}}``, ``c = \\frac{1}{\\texttt{radius}}``, and ``k = \\texttt{conic}`` .
+
+The function checks if the aspheric terms are even, odd or both and uses `EvenAsphericSurface()`,
+`OddAsphericSurface()`, or `OddEvenAsphericSurface()` as appropriate.
+
+"""
+#"pseudo-types" of aspheres
+const CONIC = 0
+const ODD = 1
+const EVEN = 2
+const ODDEVEN = 3
+
+struct AsphericSurface{T,N, Q, M} <: ParametricSurface{T,N} 
```

Comment:
```suggestion
struct AsphericSurface{T,N,Q,M<:AsphSurfaceType} <: ParametricSurface{T,N} 
```

Reply:
This does not work. If I use the @enum statement above, then CONIC etc are instances of type "AsphSurfaceType" and not individual types. For this to work, each needs to be it's own type.  I can use the @enum statement to clean this up but it must have M:<Any as the value of CONIC etc is the "type"

I find this ugly and as I get the rest fixed, I may try and see if I get similar performance by making CONIC etc subtypes of AsphSurfaceType. This would be easier to understand that what I have. 

I appreciate the help on this last round. I was stuck and now moving again. I've messed up the calculations of point and partials, still need to add stuff to benchmark and add the documentation.

Reply:
Yeah my bad, this subtype relation doesn't work